pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -184,7 +184,7 @@ public synchronized void removeConsumer(Consumer consumer) throws BrokerServiceE
                     messagesToRedeliver.add(ledgerId, entryId);
                     redeliveryTracker.addIfAbsent(PositionImpl.get(ledgerId, entryId));
                 });
-                totalAvailablePermits -= consumer.getAvailablePermits();
+                TOTAL_AVAILABLE_PERMITS_UPDATER.addAndGet(this, -consumer.getAvailablePermits());
```

Comment:
The increments/decrement on the `totalAvailablePermits` should already be safe because they're all done while holding the mutex on the dispatcher object.

Reply:
yes you are right, it seems that all updates are already happening withing `synchronized` methods. It seems that the case is similar in `NonPersistentDispatcherMultipleConsumers` where the `TOTAL_AVAILABLE_PERMITS_UPDATER` is currently used.

Reply:
@merlimat I re-opened this PR. `readMoreEntries` method isn't synchronized and it's getting called from this location without synchronization: https://github.com/apache/pulsar/blob/a1cebdb3e7cc3b8ee3ce567058182e7d31c762d2/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L2439-L2451 .